### ISSUE TITLE
Can not restart the training to obtain the same results

### ISSUE THREAD

**Issue by jyzhang-bjtu** [pytorch/pytorch#5627]
PyTorch version : 0.3.0

I use adam or sgd with the default parameters to training my network. But I found some interesing results during the training restart. Loss function is torch.nn.mse.

Algorithm sgd  : 
 First I training it with 110 batches, and save the model with torch.save(model.state_dict(), self.fn) at 100-th batch.

If I training continuously with all 110 batches, the loss I got will be the following :

2018-03-08 09:37:13 - root - INFO - SNR: 3dB Epoch[0].100  :     loss 1.2919E-03,   err_num=31756  
2018-03-08 09:39:16 - root - INFO - SNR: 3dB Epoch[0].101  :     loss 1.2755E-03,   err_num=31606   
2018-03-08 09:39:17 - root - INFO - SNR: 3dB Epoch[0].102  :     loss 1.1193E-03,   err_num=26931 
2018-03-08 09:39:18 - root - INFO - SNR: 3dB Epoch[0].103  :     loss 1.1842E-03,   err_num=28907  
2018-03-08 09:39:19 - root - INFO - SNR: 3dB Epoch[0].104  :     loss 1.3281E-03,   err_num=32834 

If I restart the training and reload the model with model.load_state_dict( torch.load(self.fn) ) at 100-th batch, then the loss I got will be the following : 

2018-03-08 09:40:51 - root - INFO - SNR: 3dB Epoch[0].100  :     loss 1.2919E-03,   err_num=31756 
2018-03-08 09:41:03 - root - INFO - SNR: 3dB Epoch[0].101  :     loss 1.2755E-03,   err_num=31606 
2018-03-08 09:41:04 - root - INFO - SNR: 3dB Epoch[0].102  :     loss 1.1193E-03,   err_num=26931 
2018-03-08 09:41:06 - root - INFO - SNR: 3dB Epoch[0].103  :     loss 1.1842E-03,   err_num=28914  
2018-03-08 09:41:07 - root - INFO - SNR: 3dB Epoch[0].104  :     loss 1.3282E-03,   err_num=32833 

for adam 
continuous training
2018-03-06 14:50:10 - root - INFO -  Epoch[0].100  :     loss 0.001053,   err_num=26686.0 
2018-03-06 14:50:11 - root - INFO -  Epoch[0].101  :     loss 0.001018,   err_num=25701.0 
2018-03-06 14:50:12 - root - INFO -  Epoch[0].102  :     loss 0.0008684,   err_num=21734.0  
2018-03-06 14:50:13 - root - INFO -  Epoch[0].103  :     loss 0.00094,   err_num=23426.0  
2018-03-06 14:50:14 - root - INFO -  Epoch[0].104  :     loss 0.001076,   err_num=26678.0  

restart training
2018-03-08 11:10:09 - root - INFO -  Epoch[0].100  :     loss 1.0526E-03,   err_num=26686  
2018-03-08 11:10:11 - root - INFO -  Epoch[0].101  :     loss 1.0135E-03,   err_num=25711  
2018-03-08 11:10:13 - root - INFO -  Epoch[0].102  :     loss 8.6998E-04,   err_num=21581  
2018-03-08 11:10:14 - root - INFO -  Epoch[0].103  :     loss 9.3137E-04,   err_num=23466  
2018-03-08 11:10:15 - root - INFO -  Epoch[0].104  :     loss 1.0608E-03,   err_num=26306  
2018-03-08 11:10:16 - root - INFO -  Epoch[0].105  :     loss 8.5917E-04,   err_num=21298 

It can be seen that loss of restart training is not same as the training with continuous training. I double check the dataset of these two case, which is all the same. Why? How did it happen?

If I use the adam alogorithm, loss and err_num will be differenet starting from Epoch[0].101.

**Comment by jyzhang-bjtu** [comment]
Then I tried to use  different dataloader and dataset, it always has the same problem.
It seems that the reloading optimizer and loss function can change the results. However, these results is different to that of continuous training.

**Comment by apaszke** [comment]
Just setting the random seed to the same value won’t let you resume the training deterministically. It’s true that the numbers generated will be the same in both runs, but you don’t start from scratch anymore, so you should skip a long prefix. A correct solution is to save RNG states.

**Comment by jyzhang-bjtu** [comment]
I use shelve to pickle the random number generator. When restart the training, I do not use seed to recover rng and use shelve to load the rng object directly. 
Then I can got the same results each time I restart the training. The script is like this.


with shelve.open('debug', r) as f0
   rng = f0['dataloader.rng']

da0 = dataloader.generate_data(rng)
model(da0)
blabla....

**Comment by apaszke** [comment]
But is data loading the only place where randomness is used? Does your model contain eg dropout?

**Comment by jyzhang-bjtu** [comment]
Yes, data loading is the only place. My model is self-defined without use any cnn and dropout, and only contains basic math operations, such as add, min, indexing, and so on.

**Comment by apaszke** [comment]
Is the data you're generating exactly the same? Maybe it's not the fault of your model. In general it's hard to figure out what's wrong with such issues unless we get a small repro, but it often is some deeply hidden mistake in the user code.

**Comment by jyzhang-bjtu** [comment]
I found that if I use algorithm-sgd without moment the results can be exact same in the first 10 epochs and will change slightly after 10 epochs.
If I can compile my model into some binary forms, I can provide my model to you for debugging.

**Comment by jyzhang-bjtu** [comment]
I will provide it later for a few months. Because I am writing a paper using pytorch, I will provide the whole thing if all the work is able to be published.

**Comment by soumith** [comment]
sounds good, I'll close the issue until then.

**Comment by wbinventor** [comment]
Was this issue ever resolved @jyzhang-bjtu? I'm running into what I think is the same problem with v0.4.1. I am trying to checkpoint my model using `torch.save(...)` and `torch.load(...)` but am unable to resume training and reproduce the same results that I observe when I continuously train. This issue occurs for the `Adam` optimizer but *not* `SGD`. The divergence in results for `Adam` occurs _after_ the first mini-batch.

For example, let's say I train for 10 iterations with a checkpoint at iteration 5. When the model is restored from the checkpoint and trained for another 5 iterations. The loss for the 1st iteration match those of the 6th iteration in the continuous run, but diverge for subsequent iterations. The losses for iterations 1-5 of the restarted run match those for iterations 6-10 of the continuous run for `SGD`, as I would expect for `Adam` or any optimizer.

I've thoroughly compared the module parameters between the continuously trained model at iteration 5 and that of the resored model and they match precisely. Is there some other portion of an `Adam` optimizers state that is not included in the state dict which precludes reproducible restarts?

I am not using dropout and have a custom dataloader with a single worker that I have confirmed produces the same data (and in the same order) when restarted from a checkpoint. I am checkpointing the random number state with `torch.get_rng_state()` and restoring it with `torch.set_rng_state()`. I am not using the native Python or NumPy RNGs, but am setting the following for reproducibility per the PyTorch docs:

```python
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```

Thanks in advance for any guidance anyone can share on how to reproducibly restore models for training!
